components/tikv_util/src/logger/file_log.rs

```diff
@@ -125,54 +153,92 @@ where
+        Self {
+            rotation_timespan,
+            next_rotation_time: None,
+        }
     }
-}
 
-impl Rotator for RotateByTime {
-    fn should_rotate(&self, file: &File) -> io::Result<bool> {
-        let metadata = file.metadata()?;
-        // Calculates how long the file has been written.
-        let file_life = metadata
-            .modified()?
-            // TODO: change it to `created()`, however, in some Linux distros it isn't supported.
-            .duration_since(metadata.accessed()?)
-            .map_err(|_| {
-                io::Error::new(
-                    ErrorKind::Other,
-                    "File accessed time is later than last-modified time",
-                )
-            })?;
-        Ok(file_life > self.rotation_timespan.0)
+    fn next_rotation_time(begin: SystemTime, duration: Duration) -> io::Result<SystemTime> {
+        begin
+            .checked_add(duration)
+            .ok_or_else(|| Error::new(ErrorKind::Other, "Next rotation time is out of range."))
     }
+}
 
+impl Rotator for RotateByTime {
     fn is_enabled(&self) -> bool {
         !self.rotation_timespan.is_zero()
     }
+
+    fn prepare(&mut self, file: &File) -> io::Result<()> {
+        self.next_rotation_time = Some(Self::next_rotation_time(
+            file.metadata()?.modified()?,
+            self.rotation_timespan.0,
+        )?);
+        Ok(())
+    }
+
+    fn should_rotate(&self) -> io::Result<bool> {
+        assert!(self.next_rotation_time.is_some());
+        Ok(Local::now() > DateTime::<Local>::from(self.next_rotation_time.unwrap()))
+    }
+
+    fn update_in_write(&mut self, _: &[u8]) -> io::Result<()> {
+        Ok(())
+    }
+
+    fn update_in_flush(&mut self) -> io::Result<()> {
+        assert!(self.next_rotation_time.is_some());
+        self.next_rotation_time = Some(Self::next_rotation_time(
+            self.next_rotation_time.unwrap(),
+            self.rotation_timespan.0,
+        )?);
+        Ok(())
+    }
 }
 
 pub struct RotateBySize {
     rotation_size: ReadableSize,
+    written: u64,
 }
 
 impl RotateBySize {
     pub fn new(rotation_size: ReadableSize) -> Self {
-        RotateBySize { rotation_size }
+        RotateBySize {
+            rotation_size,
+            written: 0,
+        }
     }
 }
 
 impl Rotator for RotateBySize {
-    fn should_rotate(&self, file: &File) -> io::Result<bool> {
-        Ok(file.metadata()?.len() > self.rotation_size.0)
-    }
-
     fn is_enabled(&self) -> bool {
         self.rotation_size.0 != 0
     }
+
+    fn prepare(&mut self, file: &File) -> io::Result<()> {
+        self.written = file.metadata()?.len();
+        Ok(())
+    }
+
+    fn should_rotate(&self) -> io::Result<bool> {
+        Ok(self.written > self.rotation_size.0)
+    }
+
+    fn update_in_write(&mut self, data: &[u8]) -> io::Result<()> {
+        self.written += data.len() as u64;
+        Ok(())
+    }
+
+    fn update_in_flush(&mut self) -> io::Result<()> {
+        Ok(())
```

Comment:
written is not cleared?

Reply:
Done

Reply:
file_size is still not updated after a rotation happens